system: Add PCI devices to hardware information

martinpitt opened the pull request:
Get them from udev's database dump. This is available on any non-ancient
Linux system and does not require privileges or new dependencies. This
also contains information about USB and SCSI devices, so these can be
added in the future without any extra calls.

Sort the table by device class by default, which appears to be the most
common view. Add click handlers to columns to change the sorting order,
in case someone wants a topological (by slot) or per-vendor list.

martinpitt commented:
Screenshot (abbreviated) from my laptop:
![PCI device list](https://user-images.githubusercontent.com/200109/35735524-72bf6678-0825-11e8-8097-1db012a37c9f.png)

martinpitt commented:
This roughly implements [Garrett's design](http://garrettlesage.com/cockpit-mockweb/hardware/big). Differences:

 * I use our standard cockpit listing
 * No "serial" as this isn't exposed in sysfs and udev, and also not really interesting IMHO.
 * "ID" is some arbitrary thing lshw makes up, it's not a generally known/recognized string. I use the PCI slot number in the first column, which is the identifier that Linux uses everywhere else. But I'm fine with hiding that and instead sorting by class (in fact this was my first version).

stefwalter commented:
@andreasn @garrett We were gonna move away from the old breadcrumb styling at the top of the page, and put that in a gray bar, that looks like the filter bar seen on other pages. Is this a good time to do this on this new page?

mareklibra commented:
LGTM, I believe this might be reused even for the virtual machine use case.

IMO, the list of devices might be quite complex, so its good to add sorting or filtering `by class`. This might help in use case of "What all network controllers are installed?"

I suggest rendering of `IOMMU Group` (if available, but quite common on servers) and `driver` as well (within a follow-up).

martinpitt commented:
@mareklibra : Right, the idea was to reuse at least the "get device list from udev". The presentation will most likely be different, though.

@garrett, @andreas: I'm not aware of our tables supporting the sorting by a particular column when clicking on them. But this might actually be useful here. WDYT?

I'm happy to change the breadcrumb to something else. Should I use the style on the docker page? I. e. gray title bar that shows  something like "Hardware Information" and a "Back to System overview" link? That would be a separate PR, as the breadcrumb is already in.

andreasn commented:
Similar to the one on the containers page?
I'm not so sure about it. Patternfly still recommends the [breadcrumb bar](http://www.patternfly.org/pattern-library/navigation/breadcrumbs/#design)  and after having used the gray bar in the containers page for a couple of months now, I feel it's not very clear.
Keeping the hardware page the way it is would align with Accounts, Storage and Networking.

stefwalter commented:
@andreasn Okay makes sense. We should however adopt the look and feel of the breadcrumb bar in Patternfly, no?

andreasn commented:
> We should however adopt the look and feel of the breadcrumb bar in Patternfly, no?

Agreed. Let's do that for this PR!
(sidenote, there seems to be two kinds of breadcrumbs in Patternfly. Filed [949](https://github.com/patternfly/patternfly/issues/949) about that )

martinpitt commented:
Thanks @andreasn , I subscribed to that PF question. Note that *this* PR does not change the breadcrumb, this needs to be changed in a separate PR as that's already in master.

andreasn commented:
Makes sense!

garrett commented:
I think the order class, model, vendor, slot might make more sense, as it prioritizes the human legibility & scanability of devices.

(It's debatable if model or vendor should go before the other. In the case of a Intel-heavy machine, like in the screenshot, vendor just adds noise. In other machines, where some parts are actually made by someone other than Intel, it's definitely important. 😉)

Having columns sortable makes sense. By default, I guess it should be sorted by slot, which will probably also happen to place devices in the same classes next to each other.

martinpitt commented:
I changed the column order as @garrett proposed:

![pci](https://user-images.githubusercontent.com/200109/35867296-ba2979ae-0b59-11e8-80a7-96f4a5e1e357.png)

I now also changed the sorting to be by device class. Like that I think it's actually more useful, as one usually is more interested in some particular device than following the (rather uninteresting and arbitrary) bus topology.

I am still looking into changing the sorting with clicking onto column headers. I don't think it's that important, but it should be possible with our `Listing` and not hard to do.

martinpitt commented:
Added clickable column sorting as well now. @garret, can you PTAL for the design? Any code reviewer?

martinpitt commented:
@mareklibra: Do you have access to a machine with IOMMU? Our openshift machines apparently don't (/sys/class/iommu/ is empty). Does that show up anywhere in `udevadm info --export-db`?
